### ISSUE TITLE
Short-Circuit with a Fat Arrow no-op flagged as fatal

### ISSUE THREAD

**Issue by twifty** [eslint/eslint#9162]
**Tell us about your environment** Atom

* **ESLint Version:** 4.5.0
* **Node Version:** 8.2.1
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?** Babel-ESLint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

{
  "extends": "eslint:recommended",
  "parser": "babel-eslint",
  "rules": {
    "strict": 0,
    "no-console": "off"
  }
}

```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js


  constructor({list, onSelection}) {
    this.items = null
    this.callbacks = {
      onSelection: onSelection || () => {}
    }
  }


```

**What did you expect to happen?**

The `onSelection: onSelection || () => {}` line is valid. It compiles and runs without errors. 

**What actually happened? Please include the actual, raw output from ESLint.**

The linter complains with an "Unexpected token (Fatal)" error.

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, this is working as intended because your code contains invalid syntax. Arrow function expressions have a lower precedence then logical expressions (expressions with `||` or `&&`), so you need to parenthesize the arrow function to make your code valid.

```js
  constructor({list, onSelection}) {
    this.items = null
    this.callbacks = {
      onSelection: onSelection || (() => {})
    }
  }
```

**Comment by twifty** [comment]
My bad. I honestly don't know why I didn't try that first.
